kimchi/src/circuits/polynomials/foreign_field_mul/circuitgates.rs

```diff
@@ -0,0 +1,384 @@
+//~
+//~ ```admonition::notice
+//~ NB: the native field modulus is obtainable from F, the native field's trait bound below.
+//~ ```
+//~
+//~ ##### Witness
+//~
+//~ * `left_input` := left foreign field element multiplicand $ ~\in F_f$
+//~ * `right_input` := right foreign field element multiplicand $ ~\in F_f$
+//~ * `quotient` := foreign field quotient $ ~\in F_f$
+//~ * `remainder` := foreign field remainder $ ~\in F_f$
+//~ * `carry0` := 2 bit carry
+//~ * `carry1_lo` := low 88 bits of `carry1`
+//~ * `carry1_hi` := high 3 bits of `carry1`
+//~ * `scaled_carry1_hi` : = `carry1_hi` scaled by 2^9
+//~ * `product1_lo` := lowest 88 bits of middle intermediate product
+//~ * `product1_hi_0` := lowest 88 bits of middle intermediate product's highest 88 + 2 bits
+//~ * `product1_hi_1` := highest 2 bits of middle intermediate product
+//~ * `quotient_bound` := quotient bound for checking `q < f`
+//~ * `quotient_bound_carry01` := quotient bound addition 1st carry bit
+//~ * `quotient_bound_carry2` := quotient bound addition 2nd carry bit
+//~
+//~ ##### Layout
+//~
+//~ The foreign field multiplication gate's rows are layed out like this
+//~
+//~ | col | `ForeignFieldMul`            | `Zero`                    |
+//~ | --- | ---------------------------- | ------------------------- |
+//~ |   0 | `left_input0`         (copy) | `remainder0`       (copy) |
+//~ |   1 | `left_input1`         (copy) | `remainder1`       (copy) |
+//~ |   2 | `left_input2`         (copy) | `remainder2`       (copy) |
+//~ |   3 | `right_input0`        (copy) | `quotient_bound01` (copy) |
+//~ |   4 | `right_input1`        (copy) | `quotient_bound2`  (copy) |
+//~ |   5 | `right_input2`        (copy) | `product1_lo`      (copy) |
+//~ |   6 | `carry1_lo`           (copy) | `product1_hi_0`    (copy) |
+//~ |   7 | `carry1_hi`        (plookup) | `product1_hi_1`           |
+//~ |   8 | `scaled_carry1_hi` (plookup) |                           |
+//~ |   9 | `carry0`                     |                           |
+//~ |  10 | `quotient0`                  |                           |
+//~ |  11 | `quotient1`                  |                           |
+//~ |  12 | `quotient2`                  |                           |
+//~ |  13 | `quotient_bound_carry01`     |                           |
+//~ |  14 | `quotient_bound_carry2`      |                           |
+//~
+
+use crate::{
+    auto_clone_array,
+    circuits::{
+        argument::{Argument, ArgumentEnv, ArgumentType},
+        expr::{constraints::ExprOps, ConstantExpr},
+        gate::GateType,
+    },
+};
+use ark_ff::PrimeField;
+use std::{array, marker::PhantomData};
+
+/// Compute non-zero intermediate products
+///
+/// For more details see the "Intermediate products" Section of
+/// the [Foreign Field Multiplication RFC](../rfcs/foreign_field_mul.md)
+///
+pub fn compute_intermediate_products<F: PrimeField, T: ExprOps<F>>(
```

Comment:
Oh, it's nice you made it to work for T, I think FFAdd works for F only 🤔

Reply:
Yeah, you can make it work for T also.